src/parsimmon.js

```diff
@@ -389,73 +399,191 @@ function formatExpected(expected) {
 }
 
 function leftPad(str, pad, char) {
-  var length = str.length;
-  if (pad <= 0) {
+  var add = pad - str.length;
+  if (add <= 0) {
     return str;
   }
-  return repeat(char, pad - length) + str;
+  return repeat(char, add) + str;
+}
+
+function toChunks(arr, chunkSize) {
+  if (!isArray(arr) || chunkSize === 0) {
+    return [[]];
+  }
+
+  var length = arr.length;
+  var chunks = [];
+  var chunkIndex = 0;
+
+  if (length <= chunkSize) {
+    return [arr.slice()];
+  }
+
+  for (var i = 0; i < length; i++) {
+    if (!chunks[chunkIndex]) {
+      chunks.push([]);
+    }
+
+    chunks[chunkIndex].push(arr[i]);
+
+    if ((i + 1) % chunkSize === 0) {
+      chunkIndex++;
+    }
+  }
+
+  return chunks;
+}
+
+function columnSum(list) {
+  if (!list || list.length === 0) {
+    return [];
+  }
+
+  var length = list.length;
+
+  if (length === 1) {
+    return list[0].slice();
+  }
+
+  var res = list[0].slice();
+  var largest = map(function(l) {
+    return l.length;
+  }, list)
+    .sort()
+    .reverse()[0];
+
+  for (var x = 0; x < largest; x++) {
+    for (var y = 1; y < length; y++) {
+      if (!res[x] || res[x] < list[y][x]) {
+        res[x] = list[y][x];
+      }
+    }
+  }
+
+  return res;
+}
+
+// Get a range of indexes including `i`-th element and `before` and `after` amount of elements from `arr`.
+function rangeFromIndexAndOffsets(i, before, after, length) {
+  // Guard against the negative upper bound for lines included in the output.
+  return {
+    from: i - before > 0 ? i - before : 0,
+    to: i + after > length ? length : i + after
+  };
+}
+
+function byteRangeToRange(byteRange) {
+  // Exception for inputs smaller than `bytesPerLine`
+  if (byteRange.from === 0 && byteRange.to === 1) {
+    return {
+      from: byteRange.from,
+      to: byteRange.to
+    };
+  }
+  return {
+    from: byteRange.from / bytesPerLine,
+    to: byteRange.to / bytesPerLine
```

Comment:
do you need to do like a `Math.floor(from/ perLine)` here so its stays an integer? seems like this could end up with weird values

Reply:
It could, thanks for pointing that out! 👍